src/main/java/net/fabricmc/loom/build/nesting/NestableJarGenerationTask.java

```diff
@@ -0,0 +1,252 @@
+
+	@InputFiles
+	@PathSensitive(PathSensitivity.NAME_ONLY)
+	protected abstract ConfigurableFileCollection getJars();
+
+	@OutputDirectory
+	public abstract DirectoryProperty getOutputDirectory();
+
+	@Input
+	protected abstract MapProperty<String, Metadata> getJarIds();
+
+	@TaskAction
+	void makeNestableJars() {
+		Map<String, String> fabricModJsons = new HashMap<>();
+		getJarIds().get().forEach((fileName, metadata) -> {
+			fabricModJsons.put(fileName, generateModForDependency(metadata));
+		});
+
+		try {
+			File targetDir = getOutputDirectory().get().getAsFile();
+			FileUtils.deleteDirectory(targetDir);
+			targetDir.mkdirs();
+		} catch (IOException e) {
+			throw new org.gradle.api.UncheckedIOException(e);
+		}
+
+		getJars().forEach(file -> {
+			File targetFile = getOutputDirectory().file(file.getName()).get().getAsFile();
+			targetFile.delete();
+			String fabricModJson = Objects.requireNonNull(fabricModJsons.get(file.getName()), "Could not generate fabric.mod.json for included dependency "+file.getName());
+			makeNestableJar(file, targetFile, fabricModJson);
+		});
+	}
+
+	public void from(Configuration configuration) {
+		ArtifactView artifacts = configuration.getIncoming().artifactView(config -> {
+			config.attributes(
+					attr -> attr.attribute(ArtifactTypeDefinition.ARTIFACT_TYPE_ATTRIBUTE, ArtifactTypeDefinition.JAR_TYPE)
+			);
+		});
+		getJars().from(artifacts.getFiles());
+		dependsOn(configuration);
+		getJarIds().set(artifacts.getArtifacts().getResolvedArtifacts().map(set -> {
+			Map<String, Metadata> map = new HashMap<>();
+			set.forEach(artifact -> {
+				ResolvedVariantResult variant = artifact.getVariant();
+
+				ComponentIdentifier id = variant.getOwner();
+				Metadata moduleLocation = null;
+
+				if (id instanceof ModuleComponentIdentifier moduleIdentifier) {
+					moduleLocation = new Metadata(
+							moduleIdentifier.getGroup(),
+							moduleIdentifier.getModule(),
+							moduleIdentifier.getVersion(),
+							null
+					);
+				}
+
+				List<Metadata> capabilityLocations = variant.getCapabilities().stream()
+						.map(capability -> new Metadata(capability.getGroup(), capability.getName(), capability.getVersion(), null))
+						.toList();
+
+				if (!capabilityLocations.isEmpty() && (moduleLocation == null || !capabilityLocations.contains(moduleLocation))) {
+					moduleLocation = capabilityLocations.get(0);
+				}
+
+				if (moduleLocation == null) {
+					throw new RuntimeException("Attempted to nest artifact " + id + " which is not a module component and has no capabilities.");
+				} else if (moduleLocation.version == null) {
+					throw new RuntimeException("Attempted to nest artifact " + id + " which has no version");
+				}
+
+				String group = moduleLocation.group;
+				String name = moduleLocation.name;
+				String version = moduleLocation.version;
+				String classifier = null;
+
+				if (artifact.getFile().getName().startsWith(name + "-" + version + "-")) {
```

Comment:
Shame there is no better way to get the classifier, this seems quite fragile.

Reply:
Yeah unfortunately this is the only way to do this (without doing the questionable approach it did before); classifiers are a maven concept that doesn't translate super well into gradle. I'd encourage people to use capability requests and feature variants instead.

All that said -- this was the workaround I found on a relevant gradle issue about the lack of a way to query artifact classifier (I can try to track that down later), and it should actually be pretty stable given that the classifier and extension provided are used only while gradle is resolving which artifact of a module to use; so if something was selected by classifier it should be in the name. At any rate, probably less fragile than the old system still.

Reply:
See https://github.com/gradle/gradle/issues/23702 -- basically, at the level we're operating at we're in variant aware resolution, and gradle doesn't really care about classifiers at that level because they're a very maven concept, so they're not there